File: pocketr-java.api/src/test/java/com/decrux/pocketr/api/services/category/ManageCategoryImplTest.java
package com.decrux.pocketr.api.services.category;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.db.ledger.CategoryTag;
import com.decrux.pocketr.api.entities.dtos.CreateCategoryDto;
import com.decrux.pocketr.api.entities.dtos.UpdateCategoryDto;
import com.decrux.pocketr.api.exceptions.ForbiddenException;
import com.decrux.pocketr.api.repositories.CategoryTagRepository;
import com.decrux.pocketr.api.services.OwnershipGuard;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.web.server.ResponseStatusException;

@DisplayName("ManageCategoryImpl")
class ManageCategoryImplTest {

    private CategoryTagRepository categoryTagRepository;
    private ManageCategoryImpl service;

    private User ownerUser;
    private User otherUser;

    @BeforeEach
    void setUp() {
        categoryTagRepository = mock(CategoryTagRepository.class);
        service = new ManageCategoryImpl(categoryTagRepository, new OwnershipGuard());
        ownerUser = createUser(1L, "alice@example.com");
        otherUser = createUser(2L, "bob@example.com");
    }

    @Nested
    @DisplayName("createCategory")
    class CreateCategory {

        @Test
        @DisplayName("should create category with unique name")
        void createCategoryWithUniqueName() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> {
                CategoryTag tag = invocation.getArgument(0);
                tag.setId(UUID.randomUUID());
                return tag;
            });

            var result = service.createCategory(new CreateCategoryDto("Groceries", null), ownerUser);
            assertEquals("Groceries", result.getName());
            assertNotNull(result.getId());
        }

        @Test
        @DisplayName("should reject duplicate category name for same user")
        void rejectDuplicateNameForSameUser() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(true);

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.createCategory(new CreateCategoryDto("Groceries", null), ownerUser)
            );
            assertEquals(409, ex.getStatusCode().value());
            assertNotNull(ex.getReason());
            assertTrue(ex.getReason().contains("already exists"));
        }

        @Test
        @DisplayName("should reject duplicate category name case-insensitively")
        void rejectDuplicateNameCaseInsensitive() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "groceries")).thenReturn(true);

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.createCategory(new CreateCategoryDto("groceries", null), ownerUser)
            );
            assertEquals(409, ex.getStatusCode().value());
            assertNotNull(ex.getReason());
            assertTrue(ex.getReason().contains("already exists"));
        }

        @Test
        @DisplayName("should allow same category name for different users")
        void allowSameNameForDifferentUsers() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false);
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(2L, "Groceries")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> {
                CategoryTag tag = invocation.getArgument(0);
                tag.setId(UUID.randomUUID());
                return tag;
            });

            var result1 = service.createCategory(new CreateCategoryDto("Groceries", null), ownerUser);
            var result2 = service.createCategory(new CreateCategoryDto("Groceries", null), otherUser);
            assertEquals("Groceries", result1.getName());
            assertEquals("Groceries", result2.getName());
        }

        @Test
        @DisplayName("should trim whitespace from category name")
        void trimWhitespaceFromName() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> {
                CategoryTag tag = invocation.getArgument(0);
                tag.setId(UUID.randomUUID());
                return tag;
            });

            var result = service.createCategory(new CreateCategoryDto("  Groceries  ", null), ownerUser);
            assertEquals("Groceries", result.getName());
        }

        @Test
        @DisplayName("should set owner to the authenticated user")
        void setOwnerToAuthenticatedUser() {
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> {
                CategoryTag tag = invocation.getArgument(0);
                assertNotNull(tag.getOwner());
                assertEquals(ownerUser.getUserId(), tag.getOwner().getUserId());
                tag.setId(UUID.randomUUID());
                return tag;
            });

            service.createCategory(new CreateCategoryDto("Food", null), ownerUser);
            verify(categoryTagRepository).save(any(CategoryTag.class));
        }
    }

    @Nested
    @DisplayName("listCategories")
    class ListCategories {

        @Test
        @DisplayName("should return all categories for owner")
        void returnAllCategoriesForOwner() {
            List<CategoryTag> tags = List.of(
                new CategoryTag(UUID.randomUUID(), ownerUser, "Groceries", null, Instant.now()),
                new CategoryTag(UUID.randomUUID(), ownerUser, "Electricity", null, Instant.now())
            );
            when(categoryTagRepository.findByOwnerUserId(1L)).thenReturn(tags);

            var result = service.listCategories(ownerUser);
            assertEquals(2, result.size());
            assertEquals("Groceries", result.get(0).getName());
            assertEquals("Electricity", result.get(1).getName());
        }

        @Test
        @DisplayName("should return empty list when user has no categories")
        void returnEmptyListWhenNoCategories() {
            when(categoryTagRepository.findByOwnerUserId(1L)).thenReturn(List.of());

            var result = service.listCategories(ownerUser);
            assertTrue(result.isEmpty());
        }
    }

    @Nested
    @DisplayName("updateCategory")
    class UpdateCategory {

        private final UUID categoryId = UUID.randomUUID();
        private CategoryTag existingTag;

        @BeforeEach
        void setUp() {
            existingTag = new CategoryTag(categoryId, ownerUser, "Groceries", null, Instant.now());
        }

        @Test
        @DisplayName("should rename category")
        void renameCategory() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> invocation.getArgument(0));

            var result = service.updateCategory(categoryId, new UpdateCategoryDto("Food", null), ownerUser);
            assertEquals("Food", result.getName());
        }

        @Test
        @DisplayName("should reject rename to existing name for same user")
        void rejectRenameToDuplicateName() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Electricity")).thenReturn(true);

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.updateCategory(categoryId, new UpdateCategoryDto("Electricity", null), ownerUser)
            );
            assertEquals(409, ex.getStatusCode().value());
        }

        @Test
        @DisplayName("should reject rename to existing name case-insensitively")
        void rejectRenameToDuplicateNameCaseInsensitive() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "electricity")).thenReturn(true);

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.updateCategory(categoryId, new UpdateCategoryDto("electricity", null), ownerUser)
            );
            assertEquals(409, ex.getStatusCode().value());
        }

        @Test
        @DisplayName("should allow renaming to the same name (no-op)")
        void allowRenamingToSameName() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> invocation.getArgument(0));

            var result = service.updateCategory(categoryId, new UpdateCategoryDto("Groceries", null), ownerUser);
            assertEquals("Groceries", result.getName());
        }

        @Test
        @DisplayName("should reject update by non-owner")
        void rejectUpdateByNonOwner() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));

            assertThrows(
                ForbiddenException.class,
                () -> service.updateCategory(categoryId, new UpdateCategoryDto("Stolen", null), otherUser)
            );
        }

        @Test
        @DisplayName("should return 404 for non-existent category")
        void notFoundForMissingCategory() {
            UUID missingId = UUID.randomUUID();
            when(categoryTagRepository.findById(missingId)).thenReturn(Optional.empty());

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.updateCategory(missingId, new UpdateCategoryDto("X", null), ownerUser)
            );
            assertEquals(404, ex.getStatusCode().value());
        }

        @Test
        @DisplayName("should trim whitespace when renaming")
        void trimWhitespaceOnRename() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));
            when(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false);
            when(categoryTagRepository.save(any(CategoryTag.class))).thenAnswer(invocation -> invocation.getArgument(0));

            var result = service.updateCategory(categoryId, new UpdateCategoryDto("  Food  ", null), ownerUser);
            assertEquals("Food", result.getName());
        }
    }

    @Nested
    @DisplayName("deleteCategory")
    class DeleteCategory {

        private final UUID categoryId = UUID.randomUUID();
        private CategoryTag existingTag;

        @BeforeEach
        void setUp() {
            existingTag = new CategoryTag(categoryId, ownerUser, "Groceries", null, Instant.now());
        }

        @Test
        @DisplayName("should delete category owned by user")
        void deleteCategoryOwnedByUser() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));

            service.deleteCategory(categoryId, ownerUser);
            verify(categoryTagRepository).delete(existingTag);
        }

        @Test
        @DisplayName("should reject delete by non-owner")
        void rejectDeleteByNonOwner() {
            when(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag));

            assertThrows(ForbiddenException.class, () -> service.deleteCategory(categoryId, otherUser));
            verify(categoryTagRepository, never()).delete(any(CategoryTag.class));
        }

        @Test
        @DisplayName("should return 404 for non-existent category")
        void notFoundForMissingCategory() {
            UUID missingId = UUID.randomUUID();
            when(categoryTagRepository.findById(missingId)).thenReturn(Optional.empty());

            ResponseStatusException ex = assertThrows(
                ResponseStatusException.class,
                () -> service.deleteCategory(missingId, ownerUser)
            );
            assertEquals(404, ex.getStatusCode().value());
        }
    }

    private static User createUser(long userId, String email) {
        User user = new User();
        user.setUserId(userId);
        user.setPassword("encoded");
        user.setEmail(email);
        return user;
    }
}

File: pocketr-api/src/test/kotlin/com/decrux/pocketr_api/services/category/ManageCategoryImplTest.kt
package com.decrux.pocketr_api.services.category

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.db.ledger.CategoryTag
import com.decrux.pocketr_api.entities.dtos.CreateCategoryDto
import com.decrux.pocketr_api.entities.dtos.UpdateCategoryDto
import com.decrux.pocketr_api.repositories.CategoryTagRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.*
import org.springframework.web.server.ResponseStatusException
import java.util.Optional
import java.util.UUID

@DisplayName("ManageCategoryImpl")
class ManageCategoryImplTest {

    private lateinit var categoryTagRepository: CategoryTagRepository
    private lateinit var service: ManageCategoryImpl

    private val ownerUser = User(
        userId = 1L,
        passwordValue = "encoded",
        email = "alice@example.com",
    )

    private val otherUser = User(
        userId = 2L,
        passwordValue = "encoded",
        email = "bob@example.com",
    )

    @BeforeEach
    fun setUp() {
        categoryTagRepository = mock(CategoryTagRepository::class.java)
        service = ManageCategoryImpl(categoryTagRepository)
    }

    @Nested
    @DisplayName("createCategory")
    inner class CreateCategory {

        @Test
        @DisplayName("should create category with unique name")
        fun createCategoryWithUniqueName() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { invocation ->
                val tag = invocation.getArgument<CategoryTag>(0)
                tag.id = UUID.randomUUID()
                tag
            }

            val result = service.createCategory(CreateCategoryDto(name = "Groceries"), ownerUser)
            assertEquals("Groceries", result.name)
            assertNotNull(result.id)
        }

        @Test
        @DisplayName("should reject duplicate category name for same user")
        fun rejectDuplicateNameForSameUser() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(true)

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.createCategory(CreateCategoryDto(name = "Groceries"), ownerUser)
            }
            assertEquals(409, ex.statusCode.value())
            assertTrue(ex.reason!!.contains("already exists"))
        }

        @Test
        @DisplayName("should reject duplicate category name case-insensitively")
        fun rejectDuplicateNameCaseInsensitive() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "groceries")).thenReturn(true)

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.createCategory(CreateCategoryDto(name = "groceries"), ownerUser)
            }
            assertEquals(409, ex.statusCode.value())
            assertTrue(ex.reason!!.contains("already exists"))
        }

        @Test
        @DisplayName("should allow same category name for different users")
        fun allowSameNameForDifferentUsers() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false)
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(2L, "Groceries")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { invocation ->
                val tag = invocation.getArgument<CategoryTag>(0)
                tag.id = UUID.randomUUID()
                tag
            }

            val result1 = service.createCategory(CreateCategoryDto(name = "Groceries"), ownerUser)
            val result2 = service.createCategory(CreateCategoryDto(name = "Groceries"), otherUser)
            assertEquals("Groceries", result1.name)
            assertEquals("Groceries", result2.name)
        }

        @Test
        @DisplayName("should trim whitespace from category name")
        fun trimWhitespaceFromName() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Groceries")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { invocation ->
                val tag = invocation.getArgument<CategoryTag>(0)
                tag.id = UUID.randomUUID()
                tag
            }

            val result = service.createCategory(CreateCategoryDto(name = "  Groceries  "), ownerUser)
            assertEquals("Groceries", result.name)
        }

        @Test
        @DisplayName("should set owner to the authenticated user")
        fun setOwnerToAuthenticatedUser() {
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { invocation ->
                val tag = invocation.getArgument<CategoryTag>(0)
                tag.id = UUID.randomUUID()
                assertEquals(ownerUser.userId, tag.owner?.userId)
                tag
            }

            service.createCategory(CreateCategoryDto(name = "Food"), ownerUser)
            verify(categoryTagRepository).save(any(CategoryTag::class.java))
        }
    }

    @Nested
    @DisplayName("listCategories")
    inner class ListCategories {

        @Test
        @DisplayName("should return all categories for owner")
        fun returnAllCategoriesForOwner() {
            val tags = listOf(
                CategoryTag(id = UUID.randomUUID(), owner = ownerUser, name = "Groceries"),
                CategoryTag(id = UUID.randomUUID(), owner = ownerUser, name = "Electricity"),
            )
            `when`(categoryTagRepository.findByOwnerUserId(1L)).thenReturn(tags)

            val result = service.listCategories(ownerUser)
            assertEquals(2, result.size)
            assertEquals("Groceries", result[0].name)
            assertEquals("Electricity", result[1].name)
        }

        @Test
        @DisplayName("should return empty list when user has no categories")
        fun returnEmptyListWhenNoCategories() {
            `when`(categoryTagRepository.findByOwnerUserId(1L)).thenReturn(emptyList())

            val result = service.listCategories(ownerUser)
            assertTrue(result.isEmpty())
        }
    }

    @Nested
    @DisplayName("updateCategory")
    inner class UpdateCategory {

        private val categoryId = UUID.randomUUID()
        private lateinit var existingTag: CategoryTag

        @BeforeEach
        fun setUp() {
            existingTag = CategoryTag(id = categoryId, owner = ownerUser, name = "Groceries")
        }

        @Test
        @DisplayName("should rename category")
        fun renameCategory() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { it.getArgument<CategoryTag>(0) }

            val result = service.updateCategory(categoryId, UpdateCategoryDto(name = "Food"), ownerUser)
            assertEquals("Food", result.name)
        }

        @Test
        @DisplayName("should reject rename to existing name for same user")
        fun rejectRenameToDuplicateName() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Electricity")).thenReturn(true)

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.updateCategory(categoryId, UpdateCategoryDto(name = "Electricity"), ownerUser)
            }
            assertEquals(409, ex.statusCode.value())
        }

        @Test
        @DisplayName("should reject rename to existing name case-insensitively")
        fun rejectRenameToDuplicateNameCaseInsensitive() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "electricity")).thenReturn(true)

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.updateCategory(categoryId, UpdateCategoryDto(name = "electricity"), ownerUser)
            }
            assertEquals(409, ex.statusCode.value())
        }

        @Test
        @DisplayName("should allow renaming to the same name (no-op)")
        fun allowRenamingToSameName() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { it.getArgument<CategoryTag>(0) }

            val result = service.updateCategory(categoryId, UpdateCategoryDto(name = "Groceries"), ownerUser)
            assertEquals("Groceries", result.name)
            // Should NOT check duplicate existence when name hasn't changed
        }

        @Test
        @DisplayName("should reject update by non-owner")
        fun rejectUpdateByNonOwner() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.updateCategory(categoryId, UpdateCategoryDto(name = "Stolen"), otherUser)
            }
            assertEquals(403, ex.statusCode.value())
        }

        @Test
        @DisplayName("should return 404 for non-existent category")
        fun notFoundForMissingCategory() {
            val missingId = UUID.randomUUID()
            `when`(categoryTagRepository.findById(missingId)).thenReturn(Optional.empty())

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.updateCategory(missingId, UpdateCategoryDto(name = "X"), ownerUser)
            }
            assertEquals(404, ex.statusCode.value())
        }

        @Test
        @DisplayName("should trim whitespace when renaming")
        fun trimWhitespaceOnRename() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))
            `when`(categoryTagRepository.existsByOwnerUserIdAndNameIgnoreCase(1L, "Food")).thenReturn(false)
            `when`(categoryTagRepository.save(any(CategoryTag::class.java))).thenAnswer { it.getArgument<CategoryTag>(0) }

            val result = service.updateCategory(categoryId, UpdateCategoryDto(name = "  Food  "), ownerUser)
            assertEquals("Food", result.name)
        }
    }

    @Nested
    @DisplayName("deleteCategory")
    inner class DeleteCategory {

        private val categoryId = UUID.randomUUID()
        private lateinit var existingTag: CategoryTag

        @BeforeEach
        fun setUp() {
            existingTag = CategoryTag(id = categoryId, owner = ownerUser, name = "Groceries")
        }

        @Test
        @DisplayName("should delete category owned by user")
        fun deleteCategoryOwnedByUser() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))

            service.deleteCategory(categoryId, ownerUser)
            verify(categoryTagRepository).delete(existingTag)
        }

        @Test
        @DisplayName("should reject delete by non-owner")
        fun rejectDeleteByNonOwner() {
            `when`(categoryTagRepository.findById(categoryId)).thenReturn(Optional.of(existingTag))

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.deleteCategory(categoryId, otherUser)
            }
            assertEquals(403, ex.statusCode.value())
            verify(categoryTagRepository, never()).delete(any())
        }

        @Test
        @DisplayName("should return 404 for non-existent category")
        fun notFoundForMissingCategory() {
            val missingId = UUID.randomUUID()
            `when`(categoryTagRepository.findById(missingId)).thenReturn(Optional.empty())

            val ex = assertThrows(ResponseStatusException::class.java) {
                service.deleteCategory(missingId, ownerUser)
            }
            assertEquals(404, ex.statusCode.value())
        }
    }
}
